src/io/ReadIO.hpp

```diff
@@ -0,0 +1,370 @@
+#pragma once
+
+#include <any>
+#include <cstdint>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <typeindex>
+#include <vector>
+
+#include <boost/multi_array.hpp>  // TODO move this and function def to the cpp file
+
+#include "Types.hpp"
+
+using StorageObjectType = AQNWB::Types::StorageObjectType;
+using SizeType = AQNWB::Types::SizeType;
+
+/*!
+ * \namespace AQNWB
+ * \brief The main namespace for AqNWB
+ */
+namespace AQNWB::IO
+{
+
+/**
+ * @brief Generic structure to hold type-erased data and shape
+ */
+class DataBlockGeneric
```

Comment:
I like the idea of only using `DataBlock` but we'll need to think a bit more about how to do that. One key difference is that `DataBlockGeneric` stores the entire data as `std::any` so we can use a single cast ``std::any_cast<std::vector<DTYPE>>(genericData.data)`` to transform the data to the correct type without having to copy the data. `DataBlock` on the other hand stores the data as and ``std::vector`` and to cast ``std::vector<std::any>`` required copying the data vector and casting each individual element. Similarly, on read we would probably also need to do another copy in order to move the data into the  ``std::vector<std::any>``.  Happy to chat about how we may be able to simplify this. 

Maybe one approach could be to have DataBlock store the data internally as ``std::any`` and then access the data via a function instead so that we do the casting when accessing the data. However, when reading data where the user may not know the data type before-hand, the user still needs to then define the type to cast to since we cannot simply cast `std::any` to `std::vector<std::any>`, but we need to cast to the correct type.